src/atomate2/aims/jobs/convergence.py

```diff
@@ -0,0 +1,189 @@
+        If the iterable is depleted and the convergence is not reached,
+        then the job is failed
+    name : str
+        A name for the job
+    maker: .BaseAimsMaker
+        A maker for the run
+    criterion_name: str
+        A name for the convergence criterion. Must be in the run results
+    epsilon: float
+        A difference in criterion value for subsequent runs
+    """
+
+    convergence_field: str
+    convergence_steps: list | tuple
+    name: str = "convergence"
+    maker: BaseAimsMaker = field(default_factory=BaseAimsMaker)
+    criterion_name: str = "energy_per_atom"
+    epsilon: float = 0.001
+
+    @job
+    def make(
+        self,
+        structure: Structure | Molecule,
+        prev_dir: str | Path | None = None,
+    ):
+        """Create a top-level flow controlling convergence iteration.
+
+        Parameters
+        ----------
+        structure : Structure or Molecule
+            a structure to run a job
+        prev_dir: str or Path or None
+            An FHI-aims calculation directory in which previous run contents are stored
+        """
+        # getting the calculation index
+        idx = 0
+        converged = False
+        if prev_dir is not None:
+            prev_dir_no_host = str(prev_dir).split(":")[-1]
+            convergence_file = Path(prev_dir_no_host) / CONVERGENCE_FILE_NAME
+            idx += 1
+            if convergence_file.exists():
+                with open(convergence_file) as f:
+                    data = json.load(f)
+                    idx = data["idx"] + 1
+                    # check for convergence
+                    converged = data["converged"]
+        else:
+            prev_dir_no_host = None
+
+        if idx < len(self.convergence_steps) and not converged:
+            # finding next jobs
+            next_base_job = self.maker.make(structure, prev_dir=prev_dir_no_host)
+            next_base_job.update_maker_kwargs(
+                {
+                    "_set": {
+                        f"input_set_generator->user_parameters->"
+                        f"{self.convergence_field}": self.convergence_steps[idx]
+                    }
+                },
+                dict_mod=True,
+            )
+            next_base_job.append_name(append_str=f" {idx}")
+
+            update_file_job = update_convergence_file(
+                prev_dir=prev_dir_no_host,
+                job_dir=next_base_job.output.dir_name,
+                criterion_name=self.criterion_name,
+                epsilon=self.epsilon,
+                convergence_field=self.convergence_field,
+                convergence_steps=self.convergence_steps,
+                output=next_base_job.output,
+            )
+
+            next_job = self.make(
+                structure,
+                prev_dir=next_base_job.output.dir_name,
+            )
+
+            replace_flow = Flow(
+                [next_base_job, update_file_job, next_job], output=next_base_job.output
```

Comment:
@ansobolev can you review this?

Reply:
This approach is nice although I'd prefer to store the current convergence information as the outputs to the last job as well, and not only the inputs of the running job - it's rather counterintuitive to find, at least for me. Also, I like files as they are independent of the DB, easy to look at and not lost when the DB gets overwritten.

Reply:
Ok, I think that should be easily accomplished just by writing the convergence data if there is a prev_dir. This also avoids the need for a separate job. I wrote up some code below:

```python
    @job
    def make(
        self,
        structure: Structure | Molecule,
        prev_dir: str | Path | None = None,
        convergence_data: dict | None = None,
        prev_output_value: float | None = None
    ):
        """Create a top-level flow controlling convergence iteration.

        Parameters
        ----------
        structure : Structure or Molecule
            a structure to run a job
        prev_dir : str or Path or None
            An FHI-aims calculation directory in which previous run contents are stored
        convergence_data : dict or None
            The convergence information to date.
        prev_output_value : float or None
            The output value being converged from the previous aims calculation.
        """
        # getting the calculation index
        idx = 0
        converged = False
        if convergence_data is not None:
            idx = convergence_data["idx"] + 1
            convergence_data["convergence_field_values"].append(self.convergence_steps[idx])
            convergence_data["criterion_values"].append(prev_output_value)
            if len(convergence_data["criterion_values"]) > 1:
                # checking for convergence
                converged = (
                    abs(prev_output_value - convergence_data["criterion_values"][-2])
                    < self.epsilon
                )
        else:
            convergence_data = {
                "criterion_name": criterion_name,
                "criterion_values": [],
                "convergence_field_name": convergence_field,
                "convergence_field_values": [],
            }
        convergence_data.update({"idx": idx, "converged": converged})
        
        if prev_dir is not None and convergence_data is not None:
            # if there is a previous directory and the convergence data is not none
            # then we have done at least one step of the convergence process and
            # we can write the convergence data in the previous directory
            split_prev_dir = str(prev_dir).split(":")[-1]
            convergence_file = Path(split_prev_dir) / CONVERGENCE_FILE_NAME
            with open(convergence_file, "w") as f:
                json.dump(convergence_data, f)

        if idx < len(self.convergence_steps) and not converged:
            # finding next jobs
            next_base_job = self.maker.make(structure, prev_dir=prev_dir)
            next_base_job.update_maker_kwargs(
                {
                    "_set": {
                        f"input_set_generator->user_parameters->"
                        f"{self.convergence_field}": self.convergence_steps[idx]
                    }
                },
                dict_mod=True,
            )
            next_base_job.append_name(append_str=f" {idx}")
            
            next_job = self.make(
                structure,
                prev_dir=next_base_job.output.dir_name,
                convergence_data=convergence_data,
                prev_output_value=getattr(next_base_job.output.output, self.criterion_name)
            )
            
            replace_flow = Flow([next_base_job, next_job], output=next_base_job.output)
            return Response(replace=replace_flow)
        
        task_doc = AimsTaskDoc.from_directory(prev_dir)
        return ConvergenceSummary.from_aims_calc_doc(task_doc)
```

Reply:
Ok, I've rewritten this. The only problem I see here now is that the index in the convergence file will be greater by 1 than the actual index of the calculation, but as this file is not actually used, I guess we can be OK with it

Reply:
Can I just clarify, in the folder with the DFT calculation, would you like the convergence file to contain the convergence info after the calculation has finished or the information without the current calculation? Both should be possible with this approach?

Reply:
Right now the file is added after the calculation has finished, which is fine with me. I don't know if the information without current calculation can be added to the current calculation directory, as it will require prior knowledge of the directory we're running in, and I'm afraid this is unavailable before running the next_base_job. Or is this directory somewhere in the attributes of the next_base_job instance? I looked into the docs, and didn't find it.